build/utils/npm.go

```diff
@@ -9,189 +9,284 @@ import (
-	// Run npm list
-	// Although this command can get --development as a flag (according to npm docs), it's not working on npm 6.
-	// Although this command can get --only=development as a flag (according to npm docs), it's not working on npm 7.
-	data, errData, err := runList(typeRestriction, executablePath, srcPath, npmArgs, log)
+// Run 'npm list ...' command and parse the returned result to create a dependencies map of.
+// The dependencies map looks like name:version -> entities.Dependency
+func CalculatePrototypeTree(executablePath, srcPath, moduleId string, npmArgs []string, cacache *cacache, log utils.Log) (map[string]*prototypeNode, error) {
+	dependenciesMap := make(map[string]*prototypeNode)
+
+	// These arguments must be added at the end of the command, to override their other values (if existed in nm.npmArgs)
+	npmArgs = append(npmArgs, "--json=true", "--all", "--long")
+	data, errData, err := RunNpmCmd(executablePath, srcPath, Ls, npmArgs, log)
 	// Some warnings and messages of npm are printed to stderr. They don't cause the command to fail, but we'd want to show them to the user.
 	if len(errData) > 0 {
 		log.Warn("Some errors occurred while collecting dependencies info:\n" + string(errData))
 	}
 	if err != nil {
-		return errors.New(fmt.Sprintf("npm list command failed with an error: %s", err.Error()))
+		log.Warn("npm list command failed with error:", err.Error())
 	}
+	npmVersion, err := GetNpmVersion(executablePath, log)
+	if err != nil {
+		return nil, err
+	}
+	parseFunc := parseNpmLsDependencyFunc(npmVersion)
 
-	// Parse the dependencies json object
-	return jsonparser.ObjectEach(data, func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) (err error) {
+	// Parse the dependencies json object.
+	return dependenciesMap, jsonparser.ObjectEach(data, func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) (err error) {
 		if string(key) == "dependencies" {
-			err = parseDependencies(value, typeRestriction, []string{moduleId}, results, log)
+			err = parseDependencies(value, []string{moduleId}, dependenciesMap, cacache, parseFunc, log)
 		}
 		return err
 	})
 }
 
+func GetNpmVersion(executablePath string, log utils.Log) (*version.Version, error) {
+	versionData, _, err := RunNpmCmd(executablePath, "", Version, nil, log)
+	if err != nil {
+		return nil, err
+	}
+	return version.NewVersion(string(versionData)), nil
+}
+
+// A partial struct of the dependency json object coming from npm ls output.
+type npmLsDependency struct {
+	Name      string
+	Version   string
+	Integrity string
+	InBundle  bool
+	Dev       bool
+	// Missing peer dependency in npm version 7/8
+	Missing bool
+	// Problems with missing peer dependency in npm version 7/8
+	Problems []string
+	// Missing  peer dependency in npm version 6
+	// Bound to 'legacyNpmLsDependency' struct
+	PeerMissing []*peerMissing
+}
+
+type legacyNpmLsDependency struct {
```

Comment:
Let's add a comment explaining this struct. 
Something like: `npm 6 ls results for a single dependency`. 
Same for the npmLsDependency struct: `npm >=7 ls results for a single dependency`.